### PR TITLE
DAMP-VSEP vocal separation using ConvTasNet

### PR DESCRIPTION
**Pull request by groadabike**
DAMPVSEP dataset 
I am using a modified version of DAMP-VSEP. The recipe handles the construction of the version.

ConvTasNet
Uses ConvTasNet - L1 loss function

Results
Only using REMIX train data. Original mixtures it is a harder problem that I am still working on.

Eval.py
I did modifications to eval.py in order to get metrics per sources and not an average of them.
This is important especially when the original mixture is used.

### CODE REVIEW

**Comment by mpariente** [comment]
Docs build fail is unrelated, but linting should be fixed before review.

**Comment by groadabike** [comment]
> Docs build fail is unrelated, but linting should be fixed before review.

Hi Manuel, can you please help me to understand the problem?

**Comment by JorisCos** [comment]
Hi! Thanks for the PR! 

It looks good to me. I think we could avoid duplicating `prepare_python_env.sh` and `parse_options.sh` by creating a symbolic link to `../../wham/ConvTasNet/utils/`.

Out of curiosity, how did you chose the hyperparameters for ConvTasNet ?

File: asteroid/data/dampvsep_dataset.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        segment (float, optional): Duration of segments in seconds,
+            Defaults to ``None`` which loads the full-length audio tracks.
+        num_workers (int, optional): Number of workers.
+            Default to ``None`` which utilise all available workers
+        norm (Str, optional): Type of normalisation to use. Default to ``None``
+            * ``'song_level'`` use mixture mean and std.
+            * ```None``` no normalisation
+        source_augmentations (Composite, optional): Default to ``None``
+        mixture (str, optional): Whether to use the original mixture with non-linear effects
+                                or remix sources. Default to original.
```
Comment (mpariente) [inline asteroid/data/dampvsep_dataset.py:57]: One tab away from `mixture`.

File: asteroid/data/dampvsep_dataset.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        self.source_augmentations = source_augmentations
+        self.mixture = mixture
+        if self.mixture == "original" and self.split == "train_english":
+            raise Exception("The 'train_english' train can only accept 'remix' mixture.")
+
+    def __len__(self):
+        return len(self.tracks) * self.samples_per_track
+
+    def _load_audio(self, path, start=0.0, duration=None, scaler=None, mean=0.0, std=1.0):
+        x, _ = librosa.load(
```
Comment (mpariente) [inline asteroid/data/dampvsep_dataset.py:111]: Import librosa here, rather than on the top. We don't have explicit deps for it.

File: egs/dampvsep/ConvTasNet/eval.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+from asteroid.data import DAMPVSEPDataset
+from asteroid.dsp import LambdaOverlapAdd
+
+
+parser = argparse.ArgumentParser()
+parser.add_argument(
+    "--out_dir",
+    type=str,
+    required=True,
+    help="Directory in exp_dir where the eval results" " will be stored",
```
Comment (mpariente) [inline egs/dampvsep/ConvTasNet/eval.py:26]: Fix please.

File: egs/dampvsep/ConvTasNet/eval.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    torch.no_grad().__enter__()
+    for idx in tqdm(range(len(test_set))):
+        # Forward the network on the mixture.
+        mix, sources = test_set[idx]
+        mix = mix.to(model_device)
+        est_sources = model.forward(mix.unsqueeze(0).unsqueeze(1))
+        mix_np = mix.squeeze(0).cpu().data.numpy()
+        sources_np = sources.cpu().data.numpy()
+        est_sources_np = est_sources.squeeze(0).cpu().data.numpy()
+        # est_sources_np = est_sources_np[[1, 0]]
```
Comment (mpariente) [inline egs/dampvsep/ConvTasNet/eval.py:85]: Remove?

File: egs/dampvsep/ConvTasNet/local/prepare_data.sh [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+python_path=python
+
+. ./utils/parse_options.sh
+if [ ! -d DAMP-VSEP-Singles ]; then
+    # Clone preprocessed DAMP-VSEP-Singles repo
+    git clone https://github.com/groadabike/DAMP-VSEP-Singles.git
+fi
+
+# Generate the splits
+. DAMP-VSEP-Singles/generate_dampvsep_singles.sh $dampvsep_root metadata $python_path
```
Comment (mpariente) [inline egs/dampvsep/ConvTasNet/local/prepare_data.sh:12]: Missing new line.

File: egs/dampvsep/ConvTasNet/run.sh [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+loss_alpha=0.5
+
+# Optim config
+optimizer=adam
+lr=0.0005
+weight_decay=0.
+
+# Data config
+task=enh_both  #'enh_vocal', 'enh_both' 
+root_path=
```
Comment (mpariente) [inline egs/dampvsep/ConvTasNet/run.sh:40]: Does this have to be overwritten?

File: egs/dampvsep/ConvTasNet/run.sh [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+	# Get ready to publish
+	mkdir -p $expdir/publish_dir
+	echo "damp-vsep/ConvTasNet" > $expdir/publish_dir/recipe_name.txt
+fi
+
+
+if [[ $stage -le 2 ]]; then
+	echo "Stage 2 : Evaluation"
+  $python_path -u eval.py --exp_dir $expdir --use_gpu 1 \
```
Comment (mpariente) [inline egs/dampvsep/ConvTasNet/run.sh:103]: Fix indent.

File: egs/dampvsep/ConvTasNet/train.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        batch_size=conf["training"]["batch_size"],
+        num_workers=conf["training"]["num_workers"],
+        drop_last=True,
+    )
+
+    val_loader = DataLoader(
+        val_set, shuffle=False, batch_size=1, num_workers=conf["training"]["num_workers"]
+    )
+
+    conf["masknet"].update({"n_src": conf["data"]["n_src"]})
```
Comment (mpariente) [inline egs/dampvsep/ConvTasNet/train.py:60]: This is like this in the other recipes because `n_src` is needed in the dataset, and can be changed inside of it. 
If `n_src` is not needed in the data, you can directly add it in conf["masknet"] from the yaml file.

File: egs/dampvsep/ConvTasNet/train.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    os.makedirs(exp_dir, exist_ok=True)
+    conf_path = os.path.join(exp_dir, "conf.yml")
+    with open(conf_path, "w") as outfile:
+        yaml.safe_dump(conf, outfile)
+
+    # Define Loss function.
+    # Combine_Loss is not complete. Needs improvement
+    # loss_func = Combine_Loss(alpha=conf['training']['loss_alpha'],
+    #                          sample_rate=conf['data']['sample_rate'])
+    loss_func = torch.nn.L1Loss()
```
Comment (mpariente) [inline egs/dampvsep/ConvTasNet/train.py:80]: Do you want to fix this before merging?

File: asteroid/data/dampvsep_dataset.py [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        1- train_english: Uses all English spoken song.
+            Duets are converted into 2 singles.
+            Totalling 9243 performances and 77Hrs.
+        2- train_singles: Uses all singles performances, discarding all duets.
+            Totalling 20660 performances and 149 hrs.
+    Args:
+        root_path (str): Root path to DAMP-VSEP dataset.
+        task (str): one of ``'enh_vocal'``,``'enh_both'``.
+            * ``'enh_vocal'`` for vocal enhanced.
+            * ``'enh_both'`` for vocal and background separation.
```
Comment (mpariente) [inline asteroid/data/dampvsep_dataset.py:38]: That's confusing for me. 
In other dataset (and in the WHAM paper), `enh_both` is used in two speaker + noise mixtures, where we want to have one output: both speakers but without the noise. Here, it's a separation task right? I think the name should be changed. `vocal_bg_sep` or simply `sep`, or something else?

File: asteroid/data/dampvsep_dataset.py [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+            dtype="float32",
+            res_type="polyphase",
+        )
+        if scaler:
+            x *= scaler
+        x -= mean
+        x /= std
+
+        if self.source_augmentations:
+            x = self.source_augmentations(x, self.sample_rate)
```
Comment (mpariente) [inline asteroid/data/dampvsep_dataset.py:126]: I guess it's ok but it seems weird to perform the augmentation in a `_load_audio` function. I'd expect just the loading to happen.

File: asteroid/data/dampvsep_dataset.py [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+# https://github.com/librosa/librosa/issues/1015
+# Soundfile can read OGG (vocal) but not M4A (background and mixture)
+import warnings
+
+warnings.filterwarnings("ignore", category=UserWarning)
+
+
+class DAMPVSEPDataset(torch.utils.data.Dataset):
+    """
+    DAMP-VSEP vocal separation dataset
```
Comment (mpariente) [inline asteroid/data/dampvsep_dataset.py:19]: I think those have to be on the same line.

File: asteroid/data/dampvsep_dataset.py [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+    Note: DAMP-VSEP dataset is hosted on Zenodo.
+          https://zenodo.org/record/3553059
+
+    Note 2: There are 2 train set available:
+        1- train_english: Uses all English spoken song.
+            Duets are converted into 2 singles.
+            Totalling 9243 performances and 77Hrs.
+        2- train_singles: Uses all singles performances, discarding all duets.
+            Totalling 20660 performances and 149 hrs.
+    Args:
```
Comment (mpariente) [inline asteroid/data/dampvsep_dataset.py:34]: New line before.

File: asteroid/data/dampvsep_dataset.py [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+
+class DAMPVSEPDataset(torch.utils.data.Dataset):
+    """
+    DAMP-VSEP vocal separation dataset
+
+    This dataset utilises one of the two preprocessed versions of DAMP-VSEP
+    from https://github.com/groadabike/DAMP-VSEP-Singles aimed for
+    SINGLE SINGER separation.
+
+    Note: DAMP-VSEP dataset is hosted on Zenodo.
```
Comment (mpariente) [inline asteroid/data/dampvsep_dataset.py:25]: Use `.. note::` instead so that it's rendered better in the docs.

**Review by mpariente — COMMENTED** [review]
Thanks ! 

Same comment as @JorisCos for the symlink (use `ln -s ../../wham/ConvTasNet/utils/ .` from the recipe folder) and some other minor comments.
In particular, the name of the task is confusing for me and should be changed.  

The readme has a lot of info, it's great !

File: asteroid/data/dampvsep_dataset.py [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+        root_path (str): Root path to DAMP-VSEP dataset.
+        task (str): one of ``'enh_vocal'``,``'enh_both'``.
+            * ``'enh_vocal'`` for vocal enhanced.
+            * ``'enh_both'`` for vocal and background separation.
+        split (str):  one of ``'train_english'``, ``'train_singles'``,
+                    ``'valid'`` and ``'test'``.
+                    Default to ``'train_singles'``.
+        samples_per_track (int, optional):
+            Number of samples yielded from each track, can be used to increase
+            dataset size, defaults to ``1``.
```
Comment (mpariente) [inline asteroid/data/dampvsep_dataset.py:44]: Maybe rename this `ex_per_track`, `samples` usually refer to audio samples in Asteroid and this could be confusing.

File: asteroid/data/dampvsep_dataset.py [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+        sample_rate=16000,
+        segment=None,
+        num_workers=None,
+        norm=None,
+        source_augmentations=None,
+        mixture="original",
+    ):
+
+        self.sample_rate = sample_rate
+        self.num_workers = cpu_count() if num_workers is None else num_workers
```
Comment (mpariente) [inline asteroid/data/dampvsep_dataset.py:80]: Is this used anywhere?

File: asteroid/data/dampvsep_dataset.py [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+
+# ignore warning related with
+# https://github.com/librosa/librosa/issues/1015
+# Soundfile can read OGG (vocal) but not M4A (background and mixture)
+import warnings
+
+warnings.filterwarnings("ignore", category=UserWarning)
+
+
+class DAMPVSEPDataset(torch.utils.data.Dataset):
```
Comment (mpariente) [inline asteroid/data/dampvsep_dataset.py:17]: Is this the `DAMPVSEPDataset` or the `DAMPVSEPSinglesDataset`?

**Review by mpariente — COMMENTED** [review]
Second round, sorry. I was having some more doubts about some things.

**Comment by groadabike** [comment]
> Out of curiosity, how did you chose the hyperparameters for ConvTasNet ?
@JorisCos I just did some experiments using a different number of blocks- repetitions and loss functions.

File: asteroid/data/dampvsep_dataset.py (hunk 16 above)
Comment (groadabike) [reply asteroid/data/dampvsep_dataset.py:80]: No, not anymore. sorry

File: asteroid/data/dampvsep_dataset.py (hunk 10 above)
Comment (groadabike) [reply asteroid/data/dampvsep_dataset.py]: I agree it is confusing. thank you

File: egs/dampvsep/ConvTasNet/train.py (hunk 9 above)
Comment (groadabike) [reply egs/dampvsep/ConvTasNet/train.py:80]: No need to wait. When using remix mixture, L1Loss works ok.
The problems are when using the original mixture.
This is why in run.sh I have 
> "mixture=remix

What do you prefer?

Comment (mpariente) [reply egs/dampvsep/ConvTasNet/train.py:78]: I let you decide about that. 
This seems fine.

Comment (groadabike) [reply egs/dampvsep/ConvTasNet/train.py:78]: Thank you @mpariente 
I think it is ready for merging. If I manage to fix the loss function, I can create another PR with a modification of the recipe.
Perhaps, one recipe that selects the loss function depending on the mixture (remix, original) or I can add a parameter to select the loss function.

Comment (mpariente) [reply egs/dampvsep/ConvTasNet/train.py:78]: I'll try to review today, thanks

File: asteroid/data/dampvsep_dataset.py [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+    def get_tracks(self):
+        """
+        Loads metadata with tracks info.
+        Creates metadata if doesn't exist.
+        """
+        metadata_path = Path(f"metadata/{self.split}_sr{self.sample_rate}.json")
+        if metadata_path.exists():
+            tracks = json.load(open(metadata_path, "r"))
+        else:
+            raise Exception(f"Metadata file for {self.split} not found")
```
Comment (mpariente) [inline asteroid/data/dampvsep_dataset.py:199]:
```suggestion
            raise RuntimeError(f"Metadata file for {self.split} not found")
```

File: egs/dampvsep/ConvTasNet/run.sh [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+
+
+	# Get ready to publish
+	mkdir -p $expdir/publish_dir
+	echo "damp-vsep/ConvTasNet" > $expdir/publish_dir/recipe_name.txt
+fi
+
+
+if [[ $stage -le 2 ]]; then
+	echo "Stage 2 : Evaluation"
```
Comment (mpariente) [inline egs/dampvsep/ConvTasNet/run.sh:100]:
```suggestion
   echo "Stage 2 : Evaluation"
```

**Review by mpariente — APPROVED** [review]
We'll just correct few things I suggested and we're good to go, thanks !

**Comment by mpariente** [comment]
Thanks !

File: asteroid/data/dampvsep_dataset.py [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+        if self.mixture == "remix":
+            audio_mix = audio_sources.sum(0)
+        else:
+            audio_mix = self._load_audio(
+                self.root_path / self.tracks[perf]["original_mix"],
+                start=start + float(self.tracks[perf]["background_start"]),
+                duration=duration,
+                mean=mix_mean,
+                std=mix_std,
+            )
```
[the remark this answers is not in the document]
Comment (jvel07) [reply asteroid/data/dampvsep_dataset.py:183]: Hi, @mpariente! :) About the augmentations, is there any recipe that does data augmentation on the fly? For example, looking at `egs/librimix/ConvTasNet` I couldn't find anything related to data augmentations (which may be used to improve a model's robustness). Is there any specific `egs` that has data augmentation in it by any chance?

**Comment by jvel07** [comment]
Hi, @mpariente! :) About the augmentations, is there any recipe that does data augmentation on the fly? For example, looking at `egs/librimix/ConvTasNet` I couldn't find anything related to data augmentations (which may be used to improve a model's robustness). Is there any specific `egs` that has data augmentation in it by any chance?

**Comment by mpariente** [comment]
This is probably what your are looking for : https://github.com/asteroid-team/asteroid/tree/654f79e41127effae7a76f48672beccf52edbf61/egs/wham/DynamicMixing

**Comment by jvel07** [comment]
@mpariente thanks for your reply. This seems to be what I am looking for, indeed. Is there any recipe that uses public corpora though?
